lib/lambda.js

```diff
@@ -2,11 +2,275 @@
 
 const constants = require('./constants')
 const shimmer = require('./instrumentation/shimmer')
+// used to detect cold starts
+let isFirstRun = true
+
+function setGenericData (transaction, event, context, isColdStart) {
+  const faas = {
+    coldstart: isColdStart,
+    execution: context.awsRequestId,
+    trigger: {
+      type: 'other'
+    }
+  }
+
+  transaction.type = 'request'
+  transaction.name = context.functionName
+  transaction.result = constants.RESULT_SUCCESS
+
+  transaction.setFaas(faas)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws'
+    }
+  }
+  transaction.setCloudContext(cloudContext)
+}
+
+function getSharedFaasData (context, isColdStart) {
+  const faas = {
+    coldstart: isColdStart,
+    execution: context.awsRequestId,
+    trigger: {
+      type: 'http'
+    }
+  }
+  return faas
+}
+
+function setApiGatewayData (transaction, event, context, isColdStart) {
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.request_id = event && event.requestContext && event.requestContext.requestId
+  faas.trigger.type = 'http'
+  transaction.setFaas(faas)
+
+  transaction.type = 'http'
+  let httpMethod = event && event.requestContext && event.requestContext.httpMethod
+  if (!httpMethod) {
+    httpMethod = event && event.requestContext && event.requestContext.http && event.requestContext.http.method
+  }
+  let path = event && event.requestContext && event.requestContext.resourcePath
+  if (!path) {
+    path = event && event.requestContext && event.requestContext.http && event.requestContext.http.path
+  }
+
```

Comment:
Also the spec says:

> The agent should use the information in the request and response objects to fill the HTTP context (context.request and context.response) fields in the same way it is done for HTTP transactions.

Perhaps a separate issue to add `transaction.context.{request,response}` fields? We won't be able to re-use the `parsers.getContextFrom{Request,Response}()` functions that take core http objects.

Reply:
Opened an issue to look into this. https://github.com/elastic/apm-agent-nodejs/issues/2419